# Python Requests 库深度指南

## 学习目标
- 掌握 requests 库的核心功能和最佳实践
- 学会处理各种 HTTP 请求场景
- 了解会话管理、认证和错误处理
- 掌握生产环境下的请求优化技巧
- 学会异步请求处理

## 前言
requests 库是 Python 中最流行的 HTTP 客户端库，被誉为"HTTP for Humans"。它简化了 HTTP 请求的复杂性，提供了优雅而强大的 API。

## 为什么选择 requests？
- **简洁的 API**：相比 urllib，更加人性化
- **功能完整**：支持所有 HTTP 方法和高级特性
- **稳定可靠**：经过大量项目验证
- **社区支持**：丰富的文档和社区资源


In [2]:
import requests
import json
from typing import Dict, Any, Optional, List
import logging
from datetime import datetime, timedelta
import time
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry
import concurrent.futures
import asyncio
import aiohttp

# 配置日志
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# 检查 requests 版本
print(f"Requests 版本: {requests.__version__}")
print(f"Python 版本兼容性: OK")
print("-" * 50)


Requests 版本: 2.32.3
Python 版本兼容性: OK
--------------------------------------------------


## 1. 基础 HTTP 请求

### 主要的 HTTP 方法
- **GET**：获取资源
- **POST**：创建资源
- **PUT**：更新资源（完整替换）
- **PATCH**：部分更新资源
- **DELETE**：删除资源
- **HEAD**：获取响应头信息
- **OPTIONS**：获取服务器支持的方法


In [3]:
# 基础请求示例

def demonstrate_basic_requests():
    """演示基础的 HTTP 请求"""
    
    print("=== 基础请求演示 ===")
    
    # 1. GET 请求
    print("\n1. GET 请求示例:")
    try:
        response = requests.get('https://httpbin.org/get')
        print(f"状态码: {response.status_code}")
        print(f"响应头: {dict(response.headers)}")
        print(f"响应内容类型: {response.headers.get('content-type')}")
        
        # 获取 JSON 数据
        data = response.json()
        print(f"请求的 URL: {data['url']}")
        print(f"服务器接收到的 Headers: {data['headers']}")
        
    except requests.exceptions.RequestException as e:
        logger.error(f"GET 请求失败: {e}")
    
    # 2. 带参数的 GET 请求
    print("\n2. 带参数的 GET 请求:")
    params = {
        'name': 'PyBackendPro',
        'course': 'Python Backend',
        'level': 'advanced'
    }
    
    try:
        response = requests.get('https://httpbin.org/get', params=params)
        data = response.json()
        print(f"请求 URL: {data['url']}")
        print(f"参数: {data['args']}")
        
    except requests.exceptions.RequestException as e:
        logger.error(f"带参数的 GET 请求失败: {e}")
    
    # 3. POST 请求
    print("\n3. POST 请求示例:")
    post_data = {
        'username': 'john_doe',
        'email': 'john@example.com',
        'action': 'register'
    }
    
    try:
        response = requests.post('https://httpbin.org/post', data=post_data)
        data = response.json()
        print(f"POST 数据: {data['form']}")
        print(f"Content-Type: {data['headers']['Content-Type']}")
        
    except requests.exceptions.RequestException as e:
        logger.error(f"POST 请求失败: {e}")

# 执行演示
demonstrate_basic_requests()


=== 基础请求演示 ===

1. GET 请求示例:
状态码: 200
响应头: {'Date': 'Mon, 16 Jun 2025 05:12:54 GMT', 'Content-Type': 'application/json', 'Content-Length': '311', 'Connection': 'keep-alive', 'Server': 'gunicorn/19.9.0', 'Access-Control-Allow-Origin': '*', 'Access-Control-Allow-Credentials': 'true'}
响应内容类型: application/json
请求的 URL: https://httpbin.org/get
服务器接收到的 Headers: {'Accept': '*/*', 'Accept-Encoding': 'gzip, deflate, br', 'Host': 'httpbin.org', 'User-Agent': 'python-requests/2.32.3', 'X-Amzn-Trace-Id': 'Root=1-684fa7d6-59d571c7242e523a16198149'}

2. 带参数的 GET 请求:
请求 URL: https://httpbin.org/get?name=PyBackendPro&course=Python+Backend&level=advanced
参数: {'course': 'Python Backend', 'level': 'advanced', 'name': 'PyBackendPro'}

3. POST 请求示例:
POST 数据: {'action': 'register', 'email': 'john@example.com', 'username': 'john_doe'}
Content-Type: application/x-www-form-urlencoded


In [ ]:

# 高级请求功能

class AdvancedRequestHandler:
    """
    高级请求处理器 - 生产环境推荐
    """
    
    def __init__(self, base_url: str = "", timeout: int = 30):
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.session = requests.Session()
        
        # 配置默认重试策略
        retry_strategy = Retry(
            total=3,  # 最大重试次数
            backoff_factor=1,  # 退避因子
            status_forcelist=[429, 500, 502, 503, 504],  # 需要重试的状态码
            allowed_methods=["HEAD", "GET", "PUT", "DELETE", "OPTIONS", "TRACE"]
        )
        
        adapter = HTTPAdapter(max_retries=retry_strategy)
        """
        这两行代码在为HTTP和HTTPS协议分别挂载重试适配器（adapter）。
        当请求失败时，这个适配器会根据之前设置的重试策略（retry_strategy）自动进行重试。
        简单来说，就是让所有HTTP和HTTPS请求都具备自动重试的能力。
        """
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)
        
        # 设置默认请求头
        """
        这里是在更新HTTP请求的默认请求头(headers)。
        update方法用于添加或修改请求头信息,具体设置了:
            User-Agent:标识客户端的名称和版本
            Accept:声明接受JSON格式的响应
            Content-Type:声明发送的数据是JSON格式
            这样之后通过这个session发送的所有请求都会带上这些请求头。
            ⚠️注意:Accept 是接收方向,Content-Type 是发送方向
        """
        self.session.headers.update({
            'User-Agent': 'PyBackendPro-Client/1.0', # 和浏览器的User-Agent不一样，Python requests 库的默认 User-Agent 就是简单的 python-requests/x.x.x，或者用户自定义一个
            'Accept': 'application/json', # Accept 表示客户端期望接收什么格式的响应数据（这里是JSON）
            'Content-Type': 'application/json' # Content-Type 表示客户端发送的请求数据的格式（这里是JSON）
        })
        
        self.logger = logging.getLogger(self.__class__.__name__)
    
    def request(self, method: str, endpoint: str, **kwargs) -> Optional[requests.Response]:
        """
        统一的请求方法
        这里是在封装一个request方法吗？核心的代码是session的request方法
            是的。这是一个封装了requests.Session.request()方法的包装器。它主要做了以下事情：
            统一处理URL拼接
            设置默认超时
            统一异常处理和日志记录
            复用session的配置（如重试策略、请求头等）
            这种封装可以让代码更加健壮和易于维护。
        Args:
            method: HTTP 方法
            endpoint: 接口端点
            **kwargs: 请求参数
            
        Returns:
            响应对象或 None
        """
        url = f"{self.base_url}/{endpoint.lstrip('/')}" if self.base_url else endpoint # 删除字符串左端所有指定的字符，这里的功能是避免出现双斜杠，所以lstrip的参数指定为斜杠，不指定的话，默认是删除左侧的所有空白字符，右侧的空白字符会得到保留
        
        # 设置默认超时
        """
        这是一种优雅的方式来处理参数的默认值：优先使用用户传入的 timeout 值，如果用户没传，就使用类中预设的 self.timeout 值。
        setdefault 是 Python 字典的一个内置方法。
        它的作用是：如果字典中不存在这个键，就设置键值对；如果已存在，则返回原有的值。
        这里是在确保 kwargs 字典中有 timeout 参数，如果没有就使用 self.timeout 作为默认值。
        这比用 if 'timeout' not in kwargs 更简洁。
        """
        kwargs.setdefault('timeout', self.timeout)
        try:
            start_time = time.time()
            response = self.session.request(method, url, **kwargs)
            end_time = time.time()
            
            self.logger.info(
                f"{method} {url} - {response.status_code} - "
                f"{end_time - start_time:.2f}s - {len(response.content)} bytes"
            )
            
            # 检查响应状态
            """
            这里在干嘛？这个方法的作用是？
                raise_for_status() 方法用于检查 HTTP 响应状态码。如果状态码是 4xx 或 5xx（表示客户端或服务器错误），它会抛出 HTTPError 异常。这是一种快速检查请求是否成功的方法。
                原因：这样可以立即捕获并处理错误的 HTTP 响应，而不是让错误的响应继续传递。
            """
            response.raise_for_status()
            return response
            
        except requests.exceptions.Timeout:
            self.logger.error(f"请求超时: {method} {url}")
        except requests.exceptions.ConnectionError:
            self.logger.error(f"连接错误: {method} {url}")
        except requests.exceptions.HTTPError as e:
            self.logger.error(f"HTTP 错误: {e}")
        except Exception as e:
            self.logger.error(f"请求失败: {e}")
        
        return None
    
    def get(self, endpoint: str, params: Dict = None) -> Optional[Dict]:
        """GET 请求的便捷方法"""
        response = self.request('GET', endpoint, params=params)
        return response.json() if response else None
    
    def post(self, endpoint: str, data: Dict = None, json_data: Dict = None) -> Optional[Dict]:
        """POST 请求的便捷方法"""
        kwargs = {}
        if json_data:
            kwargs['json'] = json_data
        elif data:
            kwargs['data'] = data
            
        response = self.request('POST', endpoint, **kwargs)
        return response.json() if response else None
    
    def close(self):
        """关闭会话"""
        self.session.close()

# 使用示例
print("=== 高级请求处理器演示 ===")
handler = AdvancedRequestHandler(base_url="https://httpbin.org")

# GET 请求
get_result = handler.get("/get", params={"test": "advanced_handler"})
if get_result:
    print(f"GET 请求成功: {get_result['args']}")

# POST 请求
post_data = {"message": "Hello from advanced handler", "timestamp": datetime.now().isoformat()}
post_result = handler.post("/post", json_data=post_data)
if post_result:
    print(f"POST 请求成功: {post_result['json']}")

# 清理资源
handler.close()


=== 高级请求处理器演示 ===


INFO:AdvancedRequestHandler:GET https://httpbin.org/get - 200 - 1.58s - 422 bytes


GET 请求成功: {'test': 'advanced_handler'}


INFO:AdvancedRequestHandler:POST https://httpbin.org/post - 200 - 0.23s - 646 bytes


POST 请求成功: {'message': 'Hello from advanced handler', 'timestamp': '2025-06-16T13:35:37.964280'}


## 2. 认证和会话管理

### 常见的认证方式
1. **Basic Authentication**：用户名密码认证
2. **Bearer Token**：令牌认证（JWT,即JSON Web Token）
    #### `Bearer Token 认证`
        Bearer Token 是一种无状态的认证方式，客户端通过在请求头中携带令牌来证明身份。
    #### 核心原理
    - Bearer = "持有者"，意思是"任何持有此令牌的人都被授权"
    - 服务器不需要存储会话状态，只需验证令牌的有效性
    - 常用于 JWT (JSON Web Token) 认证
    #### 为什么叫 "Bearer"？
    1. RFC 6750 标准：OAuth 2.0 规范中定义的术语
    2. 语义含义：Bearer = "承载者/持有者"，强调令牌本身就是凭证
    3. 安全理念：谁持有令牌，谁就有权限（类似"持票入场"）
    #### 与其他认证的区别
    ```
    Basic Auth:    Authorization: Basic dXNlcjpwYXNz
    Bearer Token:  Authorization: Bearer eyJhbGciOiJIUzI1NiI...
    API Key:       X-API-Key: your-secret-key
    ```
    - 优势：无状态、可扩展、支持微服务架构
    - 劣势：令牌泄露风险高，需要 HTTPS 传输
3. **API Key**：密钥认证
4. **OAuth**：第三方授权
5. **Session Cookies**：会话 Cookie

### 会话管理的重要性
- **连接复用**：提高性能
- **Cookie 持久化**：保持登录状态
- **请求头共享**：统一配置
- **连接池管理**：优化资源使用


In [20]:
# 认证示例

def demonstrate_authentication():
    """演示各种认证方式"""
    
    print("=== 认证方式演示 ===")
    
    # 1. Basic Authentication
    # print("\n1. Basic 认证:")
    # try:
    #     # auth时requests的get方法的参数，注意url，写死了/user/pass，所以auth里要写user和pass才能认证成功
    #     response = requests.get(
    #         'https://httpbin.org/basic-auth/user/pass',
    #         auth=('user', 'pass')
    #     )
    #     print(f"Basic 认证成功: {response.status_code}")
    #     print(f"认证信息: {response.json()}")
    # except Exception as e:
    #     logger.error(f"Basic 认证失败: {e}")
    
    # # 2. Bearer Token 认证
    # print("\n2. Bearer Token 认证:")
    # token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.example.token"
    # headers = {'Authorization': f'Bearer {token}'}
    
    # try:
    #     response = requests.get('https://httpbin.org/bearer', headers=headers)
    #     print(f"Bearer 认证状态: {response.status_code}")
    #     if response.status_code == 200:
    #         print(f"认证成功: {response.json()}")
    # except Exception as e:
    #     logger.error(f"Bearer 认证测试失败: {e}")
    
    # # 3. API Key 认证（通过请求头）
    # print("\n3. API Key 认证:")
    # api_headers = {'X-API-Key': 'your-secret-api-key-here'}
    
    # try:
    #     response = requests.get('https://httpbin.org/get', headers=api_headers)
    #     data = response.json()
    #     print(json.dumps(data,indent=2,ensure_ascii=False))
    #     print(f"API Key 已发送: {data['headers'].get('X-Api-Key', '未找到')}")
    # except Exception as e:
    #     logger.error(f"API Key 认证测试失败: {e}")
    
    # 4. 会话管理示例
    print("\n4. 会话管理:")
    session = requests.Session()
    
    # 设置会话级别的认证
    session.auth = ('session_user', 'session_pass')
    session.headers.update({'User-Agent': 'PyBackendPro-Session/1.0'}) # 用字典的方式来改其实是可以的，这里修改的是单个，所以和session.headers['User-Agent'] = 'PyBackendPro-Session/1.0'其实没有区别，需要改多个的header的配置的时候，就需要用update方法了
    
    try:
        logging.getLogger("urllib3.connectionpool").setLevel(logging.DEBUG)
        # 第一个请求
        response1 = session.get('https://httpbin.org/get')
        print(f"会话请求1: {response1.status_code}")
        print(f"第一次查看连接适配器: {session.get_adapter('https://').poolmanager.pools}")
        
        # 第二个请求（复用连接和认证）
        response2 = session.get('https://httpbin.org/get', params={'session': 'test'})
        print(f"会话请求2: {response2.status_code}")
        print(f"第二次连接适配器: {session.get_adapter('https://').poolmanager.pools}")

        response3 = requests.get('https://httpbin.org/get')
        print(f"会话请求3: {response3.status_code}")
        print(f"第三次连接适配器: {session.get_adapter('https://').poolmanager.pools}")
        
        # 会话中的 Cookie 管理
        session.cookies.set('session_id', 'abc123')
        response3 = session.get('https://httpbin.org/cookies')
        cookies_data = response3.json()
        print(f"会话 Cookies: {cookies_data['cookies']}")
        
    except Exception as e:
        logger.error(f"会话管理测试失败: {e}")
    finally:
        session.close()

# 执行认证演示
demonstrate_authentication()


DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443


=== 认证方式演示 ===

4. 会话管理:


DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /get HTTP/1.1" 200 381


会话请求1: 200
第一次查看连接适配器: <urllib3._collections.RecentlyUsedContainer object at 0x1188a3740>


DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /get?session=test HTTP/1.1" 200 419
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443


会话请求2: 200
第二次连接适配器: <urllib3._collections.RecentlyUsedContainer object at 0x1188a3740>


DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /get HTTP/1.1" 200 311


会话请求3: 200
第三次连接适配器: <urllib3._collections.RecentlyUsedContainer object at 0x1188a3740>


DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /cookies HTTP/1.1" 200 50


会话 Cookies: {'session_id': 'abc123'}


### 🤔关于会话管理的几个疑问
`1. 请问第一个请求和第二个请求我要怎么知道它是复用的？`

当前代码无法直接看出复用，你需要这样验证：
```python
# 方法1：查看连接池状态
print(f"连接适配器: {session.get_adapter('https://').poolmanager.pools}")

# 方法2：启用详细日志
import logging
logging.getLogger("urllib3.connectionpool").setLevel(logging.DEBUG)
```
- 相同域名的请求会共享TCP连接
- 认证信息（session.auth）自动携带，无需重复设置

`2. session 的 cookies 有什么用？`
主要用途：
- 状态保持：在多个请求间维持登录状态
- 会话跟踪：服务器识别是同一用户的连续操作
- 自动管理：Session会自动保存服务器返回的Set-Cookie
实际应用场景：
```python
# 登录后的API调用
session.post('/login', data={'user': 'admin', 'pass': '123'})
# 后续请求自动携带登录cookie
session.get('/dashboard')  # 无需重新认证
```
核心优势：避免每次请求都要重新认证，提升性能和用户体验。

`3. 这是urllib3的标准日志格式：方法 路径 协议版本 状态码 响应大小`

DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /get HTTP/1.1" 200 381

所以，最后的两个数字是指：响应状态码和响应的大小（单位是字节）

In [22]:

# 并发请求处理

def demonstrate_concurrent_requests():
    """演示并发请求处理"""
    
    print("=== 并发请求演示 ===")
    
    urls = [
        'https://httpbin.org/delay/1',
        'https://httpbin.org/delay/2', 
        'https://httpbin.org/delay/1',
        'https://httpbin.org/get',
        'https://httpbin.org/json'
    ]
    
    # 1. 顺序请求（慢）
    print("\n1. 顺序请求:")
    start_time = time.time()
    
    results_sequential = []
    for url in urls:
        try:
            response = requests.get(url, timeout=10)
            results_sequential.append(response.status_code)
        except Exception as e:
            results_sequential.append(f"Error: {e}")
    
    sequential_time = time.time() - start_time
    print(f"顺序请求完成: {sequential_time:.2f}s")
    print(f"结果: {results_sequential}")
    
    # 2. 并发请求（快）
    print("\n2. 并发请求:")
    start_time = time.time()
    
    def fetch_url(url):
        """获取单个 URL"""
        try:
            response = requests.get(url, timeout=10)
            return response.status_code
        except Exception as e:
            return f"Error: {e}"
    
    # 使用线程池
    with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
        results_concurrent = list(executor.map(fetch_url, urls))
    
    concurrent_time = time.time() - start_time
    print(f"并发请求完成: {concurrent_time:.2f}s")
    print(f"结果: {results_concurrent}")
    print(f"性能提升: {sequential_time/concurrent_time:.1f}x")

# 执行并发请求演示
demonstrate_concurrent_requests()


DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443


=== 并发请求演示 ===

1. 顺序请求:


DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /delay/1 HTTP/1.1" 200 361
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /delay/2 HTTP/1.1" 200 361
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /delay/1 HTTP/1.1" 200 361
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /get HTTP/1.1" 200 311
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /json HTTP/1.1" 200 429
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443
DEBUG:urllib3.connecti

顺序请求完成: 12.02s
结果: [200, 200, 200, 200, 200]

2. 并发请求:


DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /get HTTP/1.1" 200 311
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /json HTTP/1.1" 200 429
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /delay/2 HTTP/1.1" 200 361
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /delay/1 HTTP/1.1" 200 361
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /delay/1 HTTP/1.1" 200 361


并发请求完成: 3.88s
结果: [200, 200, 200, 200, 200]
性能提升: 3.1x


## 3. 生产环境最佳实践

### 性能优化技巧
1. **使用连接池**：Session 对象复用连接
2. **设置合理超时**：避免请求无限等待
3. **启用压缩**：减少网络传输
4. **批量处理**：合并多个小请求
5. **缓存策略**：缓存常用响应

### 错误处理策略
1. **分类处理**：网络错误 vs HTTP 错误
2. **重试机制**：指数退避算法
3. **熔断器**：防止雪崩效应  
4. **监控告警**：及时发现问题

### 安全考虑
1. **验证 SSL 证书**：防止中间人攻击
2. **请求头安全**：避免泄露敏感信息
3. **输入验证**：验证响应数据
4. **速率限制**：避免被限流

### 调试技巧
1. **启用详细日志**：记录请求详情
2. **使用代理**：抓包分析
3. **Mock 测试**：模拟各种场景
4. **性能分析**：监控请求耗时


In [24]:
# 生产级 HTTP 客户端封装

class ProductionHTTPClient:
    """
    生产环境 HTTP 客户端
    包含完整的错误处理、重试、监控等功能
    """
    
    def __init__(self, 
                 base_url: str = "",
                 timeout: int = 30,
                 max_retries: int = 3,
                 retry_backoff_factor: float = 0.3):
        
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.session = requests.Session()
        
        # 配置重试策略
        retry_strategy = Retry(
            total=max_retries,
            read=max_retries,
            connect=max_retries,
            backoff_factor=retry_backoff_factor,
            status_forcelist=[429, 500, 502, 503, 504],
            allowed_methods=["HEAD", "GET", "PUT", "DELETE", "OPTIONS", "TRACE"]
        )
        
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)
        
        # 默认请求头
        self.session.headers.update({
            'User-Agent': 'ProductionClient/1.0',
            'Accept': 'application/json',
            'Accept-Encoding': 'gzip, deflate',
            'Connection': 'keep-alive'
        })
        
        self.logger = logging.getLogger(self.__class__.__name__)
        
        # 统计信息
        self.stats = {
            'total_requests': 0,
            'successful_requests': 0,
            'failed_requests': 0,
            'total_time': 0.0
        }
    
    def _make_request(self, method: str, endpoint: str, **kwargs) -> Optional[requests.Response]:
        """执行请求的核心方法"""
        url = f"{self.base_url}/{endpoint.lstrip('/')}" if self.base_url else endpoint
        
        # 设置默认超时
        kwargs.setdefault('timeout', self.timeout)
        
        # 请求开始时间
        start_time = time.time()
        self.stats['total_requests'] += 1
        
        try:
            self.logger.info(f"发起请求: {method} {url}")
            
            response = self.session.request(method, url, **kwargs)
            
            # 记录耗时
            elapsed = time.time() - start_time
            self.stats['total_time'] += elapsed
            
            # 检查响应状态
            response.raise_for_status()
            
            self.stats['successful_requests'] += 1
            self.logger.info(
                f"请求成功: {method} {url} - {response.status_code} - "
                f"{elapsed:.3f}s - {len(response.content)} bytes"
            )
            
            return response
            
        except requests.exceptions.Timeout as e:
            self.stats['failed_requests'] += 1
            self.logger.error(f"请求超时: {method} {url} - {e}")
            
        except requests.exceptions.ConnectionError as e:
            self.stats['failed_requests'] += 1
            self.logger.error(f"连接错误: {method} {url} - {e}")
            
        except requests.exceptions.HTTPError as e:
            self.stats['failed_requests'] += 1
            self.logger.error(f"HTTP 错误: {method} {url} - {e}")
            
        except Exception as e:
            self.stats['failed_requests'] += 1
            self.logger.error(f"未知错误: {method} {url} - {e}")
        
        return None
    
    def get_stats(self) -> Dict[str, Any]:
        """获取客户端统计信息"""
        total = self.stats['total_requests']
        return {
            'total_requests': total,
            'successful_requests': self.stats['successful_requests'],
            'failed_requests': self.stats['failed_requests'],
            'success_rate': f"{(self.stats['successful_requests']/total*100):.1f}%" if total > 0 else "0%",
            'average_response_time': f"{(self.stats['total_time']/total):.3f}s" if total > 0 else "0s",
            'total_time': f"{self.stats['total_time']:.3f}s"
        }
    
    def get(self, endpoint: str, **kwargs) -> Optional[Dict]:
        """GET 请求"""
        response = self._make_request('GET', endpoint, **kwargs)
        try:
            return response.json() if response else None
        except ValueError:
            return {'raw_content': response.text} if response else None
    
    def post(self, endpoint: str, **kwargs) -> Optional[Dict]:
        """POST 请求"""
        response = self._make_request('POST', endpoint, **kwargs)
        try:
            return response.json() if response else None
        except ValueError:
            return {'raw_content': response.text} if response else None
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
    
    def close(self):
        """关闭客户端"""
        self.session.close()
        self.logger.info("HTTP 客户端已关闭")

# 使用示例
print("=== 生产级 HTTP 客户端演示 ===")

with ProductionHTTPClient(base_url="https://httpbin.org") as client:
    # 执行一些请求
    result1 = client.get("/get", params={"test": "production"})
    result2 = client.post("/post", json={"message": "Hello Production!"})
    result3 = client.get("/status/404")  # 这个会失败
    
    # 查看统计信息
    stats = client.get_stats()
    print(f"\n客户端统计信息:")
    for key, value in stats.items():
        print(f"  {key}: {value}")

print("\n✅ requests 库教程完成！")


INFO:ProductionHTTPClient:发起请求: GET https://httpbin.org/get
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443


=== 生产级 HTTP 客户端演示 ===


DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /get?test=production HTTP/1.1" 200 362
INFO:ProductionHTTPClient:请求成功: GET https://httpbin.org/get - 200 - 1.961s - 362 bytes
INFO:ProductionHTTPClient:发起请求: POST https://httpbin.org/post
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "POST /post HTTP/1.1" 200 524
INFO:ProductionHTTPClient:请求成功: POST https://httpbin.org/post - 200 - 0.324s - 524 bytes
INFO:ProductionHTTPClient:发起请求: GET https://httpbin.org/status/404
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /status/404 HTTP/1.1" 404 0
ERROR:ProductionHTTPClient:HTTP 错误: GET https://httpbin.org/status/404 - 404 Client Error: NOT FOUND for url: https://httpbin.org/status/404
INFO:ProductionHTTPClient:HTTP 客户端已关闭



客户端统计信息:
  total_requests: 3
  successful_requests: 2
  failed_requests: 1
  success_rate: 66.7%
  average_response_time: 0.864s
  total_time: 2.593s

✅ requests 库教程完成！
